File: HelloWorld/src/SDKDownloadLib/HttpClientUtils.cs
using System;
using System.IO;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace SDKDownloader
{
    internal static class HttpClientUtils
    {

        // Display the byte array in a readable format.
        public static string ByteArrayToString(byte[] array)
        {
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < array.Length; i++)
            {
                result.Append($"{array[i]:X2}");
            }

            return result.ToString();
        }

        private static bool CheckSha256(string file, string expectedHash)
        {
            bool result = false;

            using (SHA256 mySHA256 = SHA256.Create())
            {
                FileInfo fileInfo = new FileInfo(file);

                // Compute and print the hash values for each file in directory.
                using (FileStream fileStream = fileInfo.Open(FileMode.Open))
                {
                    try
                    {
                        // Create a fileStream for the file.
                        // Be sure it's positioned to the beginning of the stream.
                        fileStream.Position = 0;
                        // Compute the hash of the fileStream.
                        byte[] hashValue = mySHA256.ComputeHash(fileStream);
                        // Write the name and hash value of the file to the console.

                        string newHashString = ByteArrayToString(hashValue);

                        if (expectedHash == newHashString)
                        {
                            result = true;
                        }
                    }
                    catch (IOException e)
                    {
                        Console.WriteLine($"I/O Exception: {e.Message}");
                    }
                    catch (UnauthorizedAccessException e)
                    {
                        Console.WriteLine($"Access Exception: {e.Message}");
                    }
                }
            }

            return result;
        }


        public static bool DownloadValidFile(string url, string localFilePath, string sha256Hash)
        {
            bool result = false;

            HttpClientUtils.DownloadFileSynchronous(url, localFilePath);

            if (File.Exists(localFilePath))
            {
                if (sha256Hash != null)
                {

                    if (CheckSha256(localFilePath, sha256Hash))
                    {
                        result = true;
                    }
                    else
                    {
                        Console.WriteLine("Failed to validate hash : " + localFilePath);
                        Console.WriteLine("Failed to validate hash : " + sha256Hash);
                    }
                }
                else
                {
                    Logger.Log("Checksum Validation did not occur because there was no 256 key");
                    result = true;
                }

            }
            else
            {
                Console.WriteLine("Downlaod Failed URL: " + url);
            }

            return result;
        }

        public static void DownloadFileSynchronous(string url, string destPath)
        {
            using (var client = new System.Net.Http.HttpClient()) // WebClient
            {
                Logger.Log("Downloading URL: " + url);
                Logger.Log("Downloading: " + destPath);

                var fileName = destPath;
                var uri = new Uri(url);

                if (File.Exists(fileName))
                {
                    File.Delete(fileName);
                }

                Task result = client.DownloadFileTaskAsync(uri, fileName);
                result.Wait();
            }
        }



        public static async Task DownloadFileTaskAsync(this HttpClient client, Uri uri, string FileName)
        {
            using (var s = await client.GetStreamAsync(uri))
            {
                using (var fs = new FileStream(FileName, FileMode.CreateNew))
                {
                    await s.CopyToAsync(fs);
                }
            }
        }
    }

}

File: OPSWAT-SDK-Downloader/PythonSDKDownloader/HttpClientUtils.py
import os
import hashlib
import requests
import logging

class HttpClientUtils:
    @staticmethod
    def byte_array_to_string(array: bytes) -> str:
        return ''.join(f"{b:02X}" for b in array)

    @staticmethod
    def check_sha256(file_path: str, expected_hash: str) -> bool:
        try:
            with open(file_path, "rb") as f:
                sha256 = hashlib.sha256()
                for chunk in iter(lambda: f.read(4096), b""):
                    sha256.update(chunk)
                new_hash_string = HttpClientUtils.byte_array_to_string(sha256.digest())
                return expected_hash == new_hash_string
        except (IOError, PermissionError) as e:
            print(f"Exception: {e}")
            return False

    @staticmethod
    def download_valid_file(url: str, local_file_path: str, sha256_hash: str = None) -> bool:
        HttpClientUtils.download_file_synchronous(url, local_file_path)
        if os.path.exists(local_file_path):
            if sha256_hash is not None:
                if HttpClientUtils.check_sha256(local_file_path, sha256_hash):
                    return True
                else:
                    print(f"Failed to validate hash : {local_file_path}")
                    print(f"Failed to validate hash : {sha256_hash}")
                    return False
            else:
                logging.info("Checksum Validation did not occur because there was no 256 key")
                return True
        else:
            print(f"Download Failed URL: {url}")
            return False

    @staticmethod
    def download_file_synchronous(url: str, dest_path: str):
        logging.info(f"Downloading URL: {url}")
        logging.info(f"Downloading: {dest_path}")
        if os.path.exists(dest_path):
            os.remove(dest_path)
        HttpClientUtils.download_file_task(url, dest_path)

    @staticmethod
    def download_file_task(url: str, file_name: str):
        response = requests.get(url, stream=True)
        response.raise_for_status()
        with open(file_name, "wb") as f:
            for chunk in response.iter_content(chunk_size=8192):
                if chunk:
                    f.write(chunk)
